include/deal.II/lac/block_linear_operator.h

```diff
@@ -26,23 +26,185 @@
+          typename BlockMatrix>
+BlockLinearOperator<m, m, Range, Domain>
+block_diagonal_operator(const BlockMatrix &block_matrix);
+
+template <size_t m,
+          typename Range = BlockVector<double>,
+          typename Domain = Range>
+BlockLinearOperator<m, m, Range, Domain>
+block_diagonal_operator(const std::array<LinearOperator<typename Range::BlockType, typename Domain::BlockType>, m> &);
+
+template <size_t m,
+          typename Range = BlockVector<double>,
+          typename Domain = Range>
+BlockLinearOperator<m, m, Range, Domain>
+block_diagonal_operator(const LinearOperator<typename Range::BlockType, typename Domain::BlockType> &op);
+
+template <size_t n,
+          typename Range = BlockVector<double>,
+          typename Domain = Range>
+BlockLinearOperator<n, n, Domain, Range>
+block_forward_substitution(const BlockLinearOperator<n, n, Range, Domain> &,
+                           const BlockLinearOperator<n, n, Domain, Range> &);
+
+template <size_t n,
+          typename Range = BlockVector<double>,
+          typename Domain = Range>
+BlockLinearOperator<n, n, Domain, Range>
+block_back_substitution(const BlockLinearOperator<n, n, Range, Domain> &,
+                        const BlockLinearOperator<n, n, Domain, Range> &);
+
+// end of workaround
+
 /**
- * @name Creation of LinearOperator block structures
+ *  TODO: Description
+ */
+template <size_t m, size_t n, typename Range, typename Domain>
+class BlockLinearOperator : public LinearOperator<Range, Domain>
+{
+public:
+
+  typedef LinearOperator<typename Range::BlockType, typename Domain::BlockType> BlockType;
+  typedef std::array<std::array<BlockType, n>, m> ArrayType;
+
+  /**
+   * Create an empty BlockLinearOperator object. All
+   * <code>std::function</code> member objects are initialized with default
+   * variants that throw an exception upon invocation.
+   */
+  BlockLinearOperator()
+  {
+  }
+
+  /**
+   * Default copy constructor.
+   */
+  BlockLinearOperator(const BlockLinearOperator<m, n, Range, Domain> &) =
+    default;
+
+  /**
+   * Templated copy constructor that creates a BlockLinearOperator object
+   * from an object @p op for which the conversion function
+   * <code>block_operator</code> is defined.
+   */
+  template<typename Op>
+  BlockLinearOperator(const Op &op)
+  {
+    *this = block_operator<m, n, Range, Domain, Op>(op);
+  }
+
+  /**
+   * Default copy assignment operator.
+   */
+  BlockLinearOperator<m, n, Range, Domain> &
+  operator=(const BlockLinearOperator<m, n, Range, Domain> &) = default;
+
+  /**
+   * Templated copy assignment operator for an object @p op for which the
+   * conversion function <code>block_operator</code> is defined.
+   */
+  template <typename Op>
+  BlockLinearOperator<m, n, Range, Domain> &operator=(const Op &op)
+  {
+    *this = block_operator<m, n, Range, Domain, Op>(op);
+    return *this;
+  }
+
+  /**
+   * TODO: A table of the underlying linear operator blocks
+   */
+  ArrayType ops;
```

Comment:
If this is publicly accessible, a more descriptive name would be good. Maybe `blocks`?